In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'D:\Downloads\Data Analytics\content\Task 4'

In [3]:
vars_list = ['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'time_of_order', 'days_without_orders']

In [4]:
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), usecols = vars_list)

In [5]:
df_prods = pd.read_csv(os.path.join(path, 'Original', 'products.csv'), index_col = False)

In [6]:
df_test = pd.DataFrame()

In [7]:
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'))

# Task 4.5

### 4.5.1 & 4.5.2

In [22]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

From this data we can notice there seems to be an outlier on the price, as there is a max value of $99,999 which seems unusual for the data set.
We can also confirm that we cleaned some of the data, as the Count of the **product_id** is lower than the max found.


### 4.5.3

In [23]:
df_ords.dtypes

order_id                 int64
user_id                  int64
order_number             int64
orders_day_of_week       int64
time_of_order            int64
days_without_orders    float64
dtype: object

### 4.5.4

In [24]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [25]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [26]:
df_ords.dtypes

order_id                object
user_id                 object
order_number             int64
orders_day_of_week       int64
time_of_order            int64
days_without_orders    float64
dtype: object

### 4.5.5

In [27]:
df_ords.isnull().sum()

order_id                    0
user_id                     0
order_number                0
orders_day_of_week          0
time_of_order               0
days_without_orders    206209
dtype: int64

There is 206,209 lines that have missing values for **days_since_prior_order**, which is normal in this case, as those are the first purchase completed by an user, since there is no previous purchase before the first one.

### 4.5.6

In this case I will replace the missing values by 0, as this will be the day 0 since the last purchase. If removing them or creating an average the change in the **df** would be to big to be disconsidered.

In [28]:
df_ords['days_without_orders'].fillna(0, inplace=True)

In [29]:
df_ords

order_id user_id  order_number  orders_day_of_week  time_of_order  \
0        2539329       1             1                   2              8   
1        2398795       1             2                   3              7   
2         473747       1             3                   3             12   
3        2254736       1             4                   4              7   
4         431534       1             5                   4             15   
...          ...     ...           ...                 ...            ...   
3421078  2266710  206209            10                   5             18   
3421079  1854736  206209            11                   4             10   
3421080   626363  206209            12                   1             12   
3421081  2977660  206209            13                   1             12   
3421082   272231  206209            14                   6             14   

         days_without_orders  
0                        0.0  
1                       15.0  
2                       21.0  
3                       29.0  
4                       28.0  
...                      ...  
3421078                 29.0  
3421079                 30.0  
3421080                 18.0  
3421081                  7.0  
3421082                 30.0  

[3421083 rows x 6 columns]

### 4.5.7

In [30]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [31]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, time_of_order, days_without_orders]
Index: []

No duplicated values found!

### 4.5.8

If there were any, I would use ```df_ords.drop_duplicates()```

### 4.5.9

Already executed before ```df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'))```

In [32]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_checked.csv'))